runtime/src/accounts.rs

```diff
@@ -267,144 +268,183 @@ impl Accounts {
         accounts.append(&mut account_deps);
 
-        if validated_fee_payer {
-            let program_indices = message
-                .instructions()
-                .iter()
-                .map(|instruction| {
-                    self.load_executable_accounts(
-                        ancestors,
-                        &mut accounts,
-                        instruction.program_id_index as IndexOfAccount,
-                        error_counters,
-                    )
-                })
-                .collect::<Result<Vec<Vec<IndexOfAccount>>>>()?;
-
-            Ok(LoadedTransaction {
-                accounts,
-                program_indices,
-                rent: tx_rent,
-                rent_debits,
+        let program_indices = message
+            .instructions()
+            .iter()
+            .map(|instruction| {
+                let mut account_indices = Vec::new();
+                let (program_id, program_account) = accounts
+                    .get(instruction.program_id_index as usize)
+                    .ok_or(TransactionError::ProgramAccountNotFound)?;
+                let (account_found, account_dep_index) = account_found_and_dep_index
+                    .get(instruction.program_id_index as usize)
+                    .ok_or(TransactionError::ProgramAccountNotFound)?;
+                if native_loader::check_id(program_id) {
+                    return Ok(Vec::new());
+                }
+                if !account_found {
+                    error_counters.account_not_found += 1;
+                    return Err(TransactionError::ProgramAccountNotFound);
+                }
+                if !program_account.executable() {
+                    error_counters.invalid_program_for_execution += 1;
+                    return Err(TransactionError::InvalidProgramForExecution);
+                }
+                let owner_id = program_account.owner();
+                if !native_loader::check_id(owner_id) {
+                    let owner_index = if let Some(owner_index) =
+                        account_deps.iter().position(|(key, _)| key == owner_id)
+                    {
+                        accounts.len().saturating_add(owner_index)
+                    } else {
+                        let owner_index = accounts.len().saturating_add(account_deps.len());
+                        if let Some((program_account, _)) =
+                            self.accounts_db.load_with_fixed_root(ancestors, owner_id)
+                        {
+                            account_deps.push((*owner_id, program_account));
+                        } else {
+                            error_counters.account_not_found += 1;
+                            return Err(TransactionError::ProgramAccountNotFound);
+                        }
+                        owner_index
+                    };
+                    account_indices.push(owner_index as IndexOfAccount);
```

Comment:
in previous version, it calls `account_indices.insert(0, _)` to keep an order of `pd_index_n, account_index_n, ..... pd_index_0, account_index_0`. Not familiar if such order is necessary, just want to point out new version will change how account_indices is ordered.

Reply:
The loader chain was parsed in reverse and `insert(0, _)` did reverse it again so that it came out "forward". The order is correct (and stays the same as before), otherwise all subsequent tests in the program-runtime would fail with "Invalid Program" as they would try to use a custom BPF program as built-in (if the order was swapped).